### DISCUSSION TITLE
如何编译支持osx的libleaf.a？

### DISCUSSION THREAD

**Discussion by alsojin0806** [eycorsican/leaf#157] [Q&A] [answered]
查看 libleaf-ios % lipo -info libleaf.a
Architectures in the fat file: libleaf.a are: arm64 

如何编译出支持支持Mac 的lib?

**Comment by EkkoG** [comment]
make lib 即可 https://github.com/eycorsican/leaf/blob/master/Makefile#L9

**Reply by 594810899** [reply]
@EkkoG 
大佬帮忙看一下我这个代码，连不起外网，是不是哪里写的有问题，这是扩展里面的代码：

import NetworkExtension

let appGroup = "com.myncic.MyVpnProject.VPNTunneDemoA"

let conf = """
[General]
loglevel = trace
dns-server = 223.5.5.5, 114.114.114.114, 8.8.8.8
tun-fd = REPLACE-ME-WITH-THE-FD
routing-domain-resolve = true

[Proxy]
Direct = direct
Proxy = trojan, 212.50.244.81, 6666, password=myncic_123456
[Rule]
EXTERNAL, site:cn, Direct
EXTERNAL, mmdb:cn, Direct
FINAL, Proxy
"""

class PacketTunnelProvider: NEPacketTunnelProvider {
    override func startTunnel(options: [String : NSObject]?, completionHandler: @escaping (Error?) -> Void) {
        let tunnelNetworkSettings = createTunnelSettings()
        setTunnelNetworkSettings(tunnelNetworkSettings) { [weak self] error in
            let tunFd = self?.packetFlow.value(forKeyPath: "socket.fileDescriptor") as! Int32
            let confWithFd = conf.replacingOccurrences(of: "REPLACE-ME-WITH-THE-FD", with: String(tunFd))
            let url = FileManager().containerURL(forSecurityApplicationGroupIdentifier: appGroup)!.appendingPathComponent("running_config.conf")
            do {
                try confWithFd.write(to: url, atomically: false, encoding: .utf8)
            } catch {
                NSLog("fialed to write config file \(error)")
            }
            let path = url.absoluteString
            let start = path.index(path.startIndex, offsetBy: 7)
            let subpath = path[start..<path.endIndex]
            DispatchQueue.global(qos: .userInteractive).async {
                signal(SIGPIPE, SIG_IGN)
                run_leaf(String(subpath))
            }
            completionHandler(nil)
        }
    }
    
    func createTunnelSettings() -> NEPacketTunnelNetworkSettings  {

        let newSettings = NEPacketTunnelNetworkSettings(tunnelRemoteAddress: "212.50.244.81")
        newSettings.ipv4Settings = NEIPv4Settings(addresses: ["212.50.244.81"], subnetMasks: ["255.255.255.0"])
        newSettings.ipv4Settings?.includedRoutes = [NEIPv4Route.`default`()]
        newSettings.proxySettings = nil
        newSettings.dnsSettings = NEDNSSettings(servers: ["223.5.5.5", "114.114.114.114", "8.8.8.8"])
        newSettings.mtu = 1480
        
        return newSettings
    }
    
    override func stopTunnel(with reason: NEProviderStopReason, completionHandler: @escaping () -> Void) {
        completionHandler()
    }
    
    override func handleAppMessage(_ messageData: Data, completionHandler: ((Data?) -> Void)?) {
        // Add code here to handle the message.
        if let handler = completionHandler {
            handler(messageData)
        }
    }
    
    override func sleep(completionHandler: @escaping () -> Void) {
        // Add code here to get ready to sleep.
        completionHandler()
    }
    
    override func wake() {
        // Add code here to wake up.
    }
    
    
}

**Comment by alsojin0806** [comment]
tks

**Comment by alsojin0806** [comment]
@cielpy  能正常编译   iOS的规则 Mac 没办法用，就是不能正常使用，不知道是不是规则路径的原因？

osx  ASSET_LOCATION 能用不？相对路径，绝对路径都不行？

**Comment by EkkoG** [comment]
> ASSET_LOCATION

ASSET_LOCATION 可以用，https://github.com/eycorsican/leaf/issues/115

**Comment by alsojin0806** [comment]
ASSET_LOCATION  写配置信息里面吗？我各种尝试都不行

**Comment by EkkoG** [comment]
> ASSET_LOCATION 写配置信息里面吗？我各种尝试都不行

声明一个环境变量 https://github.com/eycorsican/leaf/issues/115#issuecomment-815586072

**Comment by alsojin0806** [comment]
不知道这个环境变量设置哪里路径比较合理，尝试归档，group ,本地静态路径都不行，要不就是没得权限，链接成功后，代理没成功.
查看这边，貌似是获取`ASSET_LOCATION (https://github.com/eycorsican/leaf/blob/master/leaf/src/option/mod.rs#L58) 对应的环境变量值

**Comment by alsojin0806** [comment]
用leaf_test_config 测试相对路径规则 返回的是的是0，在传给leaf_run，如果设置tun-fd 等于具体某一个值，直接挂了，好像也是返回0，如果不设置这个值，不出现问题，但是代理没法用，应该不是路径问题？

**Comment by EkkoG** [comment]
如果你用 Swift，调用 `leaf_run` 前调用 `setenv` 设置环境变量
```
setenv("ASSET_LOCATION", "/Users/xxx/xxx/", 1)
```

然后把资源文件放到 /Users/xxx/xxx/ 下就可以了

添加一条规则

```
EXTERNAL,site:category-ads-all,Reject
```

启动 leaf 日志里会有显示加载成功

```
loaded 639 domain rules from [/Users/xxx/xxx/site.dat] for tag [category-ads-all]
```

正常情况下，整个流程就是这样的了

**Comment by alsojin0806** [comment]
不行  已经测试过了,我猜你用的是单机版
我是用的插件ne，这地方貌似不能直接用ne,tun-fd 这边貌似有点问题，如果用ne
![image](https://user-images.githubusercontent.com/3974508/115574291-13a17380-a2f4-11eb-8884-9c4c218ef7ce.png)

**Comment by EkkoG** [comment]
我实测，是可以的

**Comment by alsojin0806** [comment]
> 我实测，是可以的

你用的是NE吗？

**Comment by EkkoG** [comment]
> 不行 已经测试过了,我猜你用的是单机版
> 我是用的插件ne，这地方貌似不能直接用ne,tun-fd 这边貌似有点问题，如果用ne
> ![image](https://user-images.githubusercontent.com/3974508/115574291-13a17380-a2f4-11eb-8884-9c4c218ef7ce.png)

什么是单机版？

我没有用 NE，直接调用 leaf_run 运行的

用 NE leaf 应该是运行在另一个进程吧，setenv 应该是作用于当前进程

**Comment by alsojin0806** [comment]
> > 不行 已经测试过了,我猜你用的是单机版
> > 我是用的插件ne，这地方貌似不能直接用ne,tun-fd 这边貌似有点问题，如果用ne
> > ![image](https://user-images.githubusercontent.com/3974508/115574291-13a17380-a2f4-11eb-8884-9c4c218ef7ce.png)
> 
> 什么是单机版？
> 
> 我没有用 NE，直接调用 leaf_run 运行的
> 
> 用 NE leaf 应该是运行在另一个进程吧，setenv 应该是作用于当前进程

明白了 单机版是指那种不用通过苹果商店奋发的

**Comment by alsojin0806** [comment]
> 如果你用 Swift，调用 `leaf_run` 前调用 `setenv` 设置环境变量
> 
> ```
> setenv("ASSET_LOCATION", "/Users/xxx/xxx/", 1)
> ```
> 
> 然后把资源文件放到 /Users/xxx/xxx/ 下就可以了
> 
> 添加一条规则
> 
> ```
> EXTERNAL,site:category-ads-all,Reject
> ```
> 
> 启动 leaf 日志里会有显示加载成功
> 
> ```
> loaded 639 domain rules from [/Users/xxx/xxx/site.dat] for tag [category-ads-all]
> ```
> 
> 正常情况下，整个流程就是这样的了

 刚才测试了一下是这样的，用NE的话，设置本地路径 没得权限的，还发现一个问题，传入配置文件的路径中如果含有空格，传入leaf_run 会导致没办法找到相应的路径的，不知道这位开源大神有没有遇到过

**Comment by EkkoG** [comment]
> > 如果你用 Swift，调用 `leaf_run` 前调用 `setenv` 设置环境变量
> > ```
> > setenv("ASSET_LOCATION", "/Users/xxx/xxx/", 1)
> > ```
> > 
> > 
> > 然后把资源文件放到 /Users/xxx/xxx/ 下就可以了
> > 添加一条规则
> > ```
> > EXTERNAL,site:category-ads-all,Reject
> > ```
> > 
> > 
> > 启动 leaf 日志里会有显示加载成功
> > ```
> > loaded 639 domain rules from [/Users/xxx/xxx/site.dat] for tag [category-ads-all]
> > ```
> > 
> > 
> > 正常情况下，整个流程就是这样的了
> 
> 刚才测试了一下是这样的，用NE的话，设置本地路径 没得权限的，还发现一个问题，传入配置文件的路径中如果含有空格，传入leaf_run 会导致没办法找到相应的路径的，不知道这位开源大神有没有遇到过

我没有搞过 NE 的开发，猜测 NE 应该是一个受限的沙盒进程，所以不能访问本地路径吧

没有遇到过，macOS 的路径有空格时需要处理吧，个人是尽量避免路径有空格的，所以没遇到

**Comment by alsojin0806** [comment]
start with options:
StartOptions {
    config: File(
        "/Users/xxxxxx/Library/Group Containers/xxxxxxtest/config.conf",
    ),
    runtime_opt: SingleThread,
}
loaded 629 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [category-ads-all]
loaded 2665 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [cn]
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" })', [leaf/src/proxy/tun/inbound.rs:88:42](https://github.com/eycorsican/leaf/blob/master/leaf/src/proxy/tun/inbound.rs#L88)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

有遇到这种问题没，用NE 用到tun-fd 就挂了

**Reply by zhimeilinzm** [reply]
你解决了吗

**Comment by EkkoG** [comment]
> start with options:
> StartOptions {
> config: File(
> "/Users/xxxxxx/Library/Group Containers/xxxxxxtest/config.conf",
> ),
> runtime_opt: SingleThread,
> }
> loaded 629 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [category-ads-all]
> loaded 2665 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [cn]
> thread '' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" })', [leaf/src/proxy/tun/inbound.rs:88:42](https://github.com/eycorsican/leaf/blob/master/leaf/src/proxy/tun/inbound.rs#L88)
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 
> 有遇到这种问题没，用NE 用到tun-fd 就挂了

@wangjinwei0806 没有权限，日志里说的很清楚了

**Comment by alsojin0806** [comment]
> > start with options:
> > StartOptions {
> > config: File(
> > "/Users/xxxxxx/Library/Group Containers/xxxxxxtest/config.conf",
> > ),
> > runtime_opt: SingleThread,
> > }
> > loaded 629 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [category-ads-all]
> > loaded 2665 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [cn]
> > thread '' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" })', [leaf/src/proxy/tun/inbound.rs:88:42](https://github.com/eycorsican/leaf/blob/master/leaf/src/proxy/tun/inbound.rs#L88)
> > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > 有遇到这种问题没，用NE 用到tun-fd 就挂了
> 
> @wangjinwei0806 没有权限，日志里说的很清楚了

我知道是没得权限，但是leaflib 这个框架怎么给权限，这个需要如何操作，报错这个不是用的NE，按照你上面描述的设置setenv 环境变量，后面直接出现上面这个错误了，是不是使用方式不对，“loaded 629 domain rules ”有这些日志了，説名路径正常

**Comment by EkkoG** [comment]
> > > start with options:
> > > StartOptions {
> > > config: File(
> > > "/Users/xxxxxx/Library/Group Containers/xxxxxxtest/config.conf",
> > > ),
> > > runtime_opt: SingleThread,
> > > }
> > > loaded 629 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [category-ads-all]
> > > loaded 2665 domain rules from [/Users/xxxxxx/Library/Group Containers/xxxxxxtest/site.dat] for tag [cn]
> > > thread '' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" })', [leaf/src/proxy/tun/inbound.rs:88:42](https://github.com/eycorsican/leaf/blob/master/leaf/src/proxy/tun/inbound.rs#L88)
> > > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > > 有遇到这种问题没，用NE 用到tun-fd 就挂了
> > 
> > 
> > @wangjinwei0806 没有权限，日志里说的很清楚了
> 
> 我知道是没得权限，但是leaflib 这个框架怎么给权限，这个需要如何操作，报错这个不是用的NE，按照你上面描述的设置setenv 环境变量，后面直接出现上面这个错误了，是不是使用方式不对，“loaded 629 domain rules ”有这些日志了，説名路径正常

这个需要你自己结合 NE 文档查了，我并不清楚，抱歉

**Comment by alsojin0806** [comment]
上面的没用NE，你在运行lib的时候有遇到以下日志吗？
start with options:
StartOptions {
    config: File(
        "/Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/config.conf",
    ),
    runtime_opt: SingleThread,
}
loaded 629 domain rules from [/Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/site.dat] for tag [category-ads-all]
loaded 2665 domain rules from [/Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/site.dat] for tag [cn]
[33m[2021-04-22 18:18:45][[33mWARN[0m[33m] open mmdb file /Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/mmdb failed[0m
[33m[2021-04-22 18:18:45][[33mWARN[0m[33m] empty rule at target Direct[0m
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" })', leaf/src/proxy/tun/inbound.rs:88:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by EkkoG** [comment]
> 上面的没用NE，你在运行lib的时候有遇到以下日志吗？
> 
> start with options:
> 
> StartOptions {
> 
>     config: File(
> 
>         "/Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/config.conf",
> 
>     ),
> 
>     runtime_opt: SingleThread,
> 
> }
> 
> loaded 629 domain rules from [/Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/site.dat] for tag [category-ads-all]
> 
> loaded 2665 domain rules from [/Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/site.dat] for tag [cn]
> 
> [33m[2021-04-22 18:18:45][[33mWARN[0m[33m] open mmdb file /Users/xxxxxxx/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/mmdb failed[0m
> 
> [33m[2021-04-22 18:18:45][[33mWARN[0m[33m] empty rule at target Direct[0m
> 
> thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" })', leaf/src/proxy/tun/inbound.rs:88:42
> 
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Xcode scheme 里，调整成用root用户运行

**Comment by alsojin0806** [comment]
start with options:
StartOptions {
    config: File(
        "/var/root/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/config.conf",
    ),
    runtime_opt: SingleThread,
}
loaded 629 domain rules from [/var/root/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/site.dat] for tag [category-ads-all]
loaded 2665 domain rules from [/var/root/Library/Developer/Xcode/DerivedData/testleaf-ewmwyppsbjyxhrbrzvhgllacauzj/Build/Products/Debug/testleaf.app/Contents/Resources/site.dat] for tag [cn]
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 16, kind: Other, message: "Resource busy" })', leaf/src/proxy/tun/inbound.rs:88:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

"Resource busy" 资源被占用，我重启电脑还是这样

**Comment by EkkoG** [comment]
是不是有另一个 leaf 进程在跑而且用的相同的配置文件？资源有竞争了？

**Reply by alsojin0806** [reply]
后面我重启电脑了，还是类似问题，我在试试看

**Comment by eycorsican** [comment]
提供一下完整调用代码和 配置

**Reply by alsojin0806** [reply]
就几行简单代码 我本身是测试路径的
![image](https://user-images.githubusercontent.com/3974508/115742891-b2e16c00-a3c3-11eb-8c4c-f6b2b20f6573.png)

**Reply by alsojin0806** [reply]
[General]
loglevel = info
dns-server = 223.5.5.5, 114.114.114.114
interface = 127.0.0.1
port = 1087
socks-interface = 127.0.0.1
socks-port = 1086
tun-fd = 4
[Proxy]
Direct = direct
Reject = reject
vmess-out = vmess, xxxx, 443, username=xxxx, ws=true, tls=true, ws-path=/xxxx
[Proxy Group]
Fallback = fallback, vmess-out, interval=600, timeout=5
[Rule]
EXTERNAL, site:category-ads-all, Reject
EXTERNAL, site:cn, Direct
EXTERNAL, mmdb:cn, Direct
DOMAIN-SUFFIX,0rz.tw,Fallback
FINAL, Fallback

**Reply by eycorsican** [reply]
tun-fd 接收的是一个已经打开了的 fd

**Reply by EkkoG** [reply]
我遇到 Resource busy 错误的场景是因为另一个 leaf 进程在运行

```
2021-04-22 23:52:46.413420+0800 LeafX[12476:71892] \^[[37m[2021-04-22 23:52:46][\^[[37mDEBUG\^[[0m\^[[37m] default handler [Proxy]\^[[0m
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 16, kind: Other, message: "Resource busy" })', leaf/src/proxy/tun/inbound.rs:88:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2021-04-22 23:52:46.417469+0800 LeafX[12476:71890] done
```

杀了另一个进程就可以了

**Reply by alsojin0806** [reply]
> tun-fd 接收的是一个已经打开了的 fd
fd 怎么配置的，有规则没？

**Reply by alsojin0806** [reply]
> 我遇到 Resource busy 错误的场景是因为另一个 leaf 进程在运行
> 
> ```
> 2021-04-22 23:52:46.413420+0800 LeafX[12476:71892] \^[[37m[2021-04-22 23:52:46][\^[[37mDEBUG\^[[0m\^[[37m] default handler [Proxy]\^[[0m
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 16, kind: Other, message: "Resource busy" })', leaf/src/proxy/tun/inbound.rs:88:42
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 2021-04-22 23:52:46.417469+0800 LeafX[12476:71890] done
> ```
> 
> 杀了另一个进程就可以了

我尝试电脑重启 再次运行 还是这样的问题
日志如下
.............
loaded 629 domain rules from [/Users/xxx/Desktop/22222/site.dat] for tag [category-ads-all]
loaded 2665 domain rules from [/Users/xxx/Desktop/22222/site.dat] for tag [cn]
[33m[2021-04-23 00:10:52][[33mWARN[0m[33m] open mmdb file /Users/wangjinwei/Desktop/22222/geo.mmdb failed[0m
[33m[2021-04-23 00:10:52][[33mWARN[0m[33m] empty rule at target Direct[0m
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 16, kind: Other, message: "Resource busy" })', leaf/src/proxy/tun/inbound.rs:88:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Reply by alsojin0806** [reply]
![image](https://user-images.githubusercontent.com/3974508/115749374-9c3e1380-a3c9-11eb-9763-6b9383545218.png)

**Reply by alsojin0806** [reply]
> tun-fd 接收的是一个已经打开了的 fd

看见这个了
https://github.com/eycorsican/rust-tun#macos

**Reply by eycorsican** [reply, marked as answer]
最后所需的代码也已经补上了：https://github.com/eycorsican/rust-tun/commit/1ada024af12bc208c4bdc87070e58fe618e694e7

编译可能要 cargo update 一下

> fd 怎么配置的，有规则没？

NE 框架，跟 iOS 的应该没什么区别

**Reply by azhm** [reply]
@wangjinwei0806 你截图这种我测试是没问题  但是我通过NE创建Tunnel 会无法启用

![image](https://user-images.githubusercontent.com/12508146/115764620-5d638a00-a3d8-11eb-91d0-70b9acef06ef.png)

**Reply by eycorsican** [reply]
https://github.com/eycorsican/xleaf

**Reply by alsojin0806** [reply]
> 最后所需的代码也已经补上了：[eycorsican/rust-tun@1ada024](https://github.com/eycorsican/rust-tun/commit/1ada024af12bc208c4bdc87070e58fe618e694e7)
> 
> 编译可能要 cargo update 一下
> 
> > fd 怎么配置的，有规则没？
> 
> NE 框架，跟 iOS 的应该没什么区别
test success

**Reply by alsojin0806** [reply]
> 你截图这种我测试是没问题 但是我通过NE创建Tunnel 会无法启用
@azhm
> 
> ![image](https://user-images.githubusercontent.com/12508146/115764620-5d638a00-a3d8-11eb-91d0-70b9acef06ef.png)

不知道是不是你编译lib的时候，作者已经修复了这个问题 
https://github.com/eycorsican/rust-tun/commit/1ada024af12bc208c4bdc87070e58fe618e694e7

**Reply by alsojin0806** [reply]
> https://github.com/eycorsican/xleaf
貌似看不到日志，osx 用NE插件，插件看不到日志
